### PR TITLE
Files, Events Refactored

### PR DESCRIPTION
**Pull request by TangyrenMidKing**
Event:
cancel_register/{event_id}
change_status
create_event
delete_event​/{event_id}
get_event/{event_id}
get_next_event_page
get_recent_events
register_event/{event_id}
update_event

Files:
change_status
delete_file/{file_id}
get_file/{file_id}
update_file
upload_file

### CODE REVIEW

File: src/api/events.py [hunk 1]
```diff
@@ -95,44 +103,46 @@ def delete_event(event_id: str) -> Response:
...
+    # fetch event data from firestore
+    event_ref = db.collection("Scheduled-Events").document(event_id)
+    event = event_ref.get().to_dict()
 
-        # Delete document
-        for doc in docs:
-            # Store document id and use it to locate the specific document to delete
-            doc_id = doc.id
-            db.collection("Scheduled-Events").document(doc_id).delete()
+    # if enevt does not exists
```
Comment (Astatium5) [inline src/api/events.py:121]: "event"

File: firestore.rules [hunk 2]
```diff
@@ -1,18 +1,21 @@
...
-      allow read: if request.auth != null;
-    }
-    match /Roles/allRoles {
-    // create and assign roles to users, only for Admins
-    allow read, write: if request.auth.token.admin == true;
-    }
+    
+    allow read, write: if request.auth != null;
+    
+		// match /Scheduled-Events/{events} {
```
Comment (Astatium5) [inline firestore.rules:7]: Remove comments

File: src/api/events.py [hunk 3]
```diff
@@ -143,55 +153,55 @@ def update_event() -> Response:
...
+    event_ref = db.collection("Scheduled-Events").document(event_id)
+    event = event_ref.get().to_dict()
 
-        docs: list = (
-            db.collection("Scheduled-Events")
-            .where("id", "==", event_id)
-            .where("author", "==", uid)
-            .stream()
-        )
+    # if enevt does not exists
```
Comment (Astatium5) [inline src/api/events.py:179]: "event"

File: src/api/events.py [hunk 4]
```diff
@@ -342,3 +358,168 @@ def get_next_event_page() -> Response:
...
+    # fetch event data from firestore
+    event_ref = db.collection("Scheduled-Events").document(event_id)
+    event = event_ref.get().to_dict()
+
+    # if enevt does not exists
+    if not event_ref.get().exists:
+        return Response("The event no longer exists", 404)
+    # Only the event organisor or the admin could delete the event
+    if event["status"] > 1:
+        return Response("The event close for register", 403)
```
Comment (Astatium5) [inline src/api/events.py:406]: "is closed for registration"
Also, it should return a `Forbidden` exception imported from `werkzeug.exceptions`

File: src/api/events.py [hunk 5]
```diff
@@ -342,3 +358,168 @@ def get_next_event_page() -> Response:
...
+    decoded_token: dict = auth.verify_id_token(token)
+    uid: str = decoded_token["uid"]
+
+    # fetch event data from firestore
+    event_ref = db.collection("Scheduled-Events").document(event_id)
+    event = event_ref.get().to_dict()
+
+    # if enevt does not exists
+    if not event_ref.get().exists:
+        return Response("The event no longer exists", 404)
```
Comment (Astatium5) [inline src/api/events.py:403]: Should be `NotFound` from `werkzeug.exceptions`

File: src/api/events.py [hunk 6]
```diff
@@ -342,3 +358,168 @@ def get_next_event_page() -> Response:
...
```
Comment (Astatium5) [inline src/api/events.py:458]: Should be `NotFound` from `werkzeug.exceptions`

File: src/api/events.py [hunk 7]
```diff
@@ -342,3 +358,168 @@ def get_next_event_page() -> Response:
...
+    """
+    # check tokens and get uid from token
+    token: str = request.headers["Authorization"]
+    decoded_token: dict = auth.verify_id_token(token)
+    reviewer: str = decoded_token["uid"]
+    data: dict = request.get_json()
+
+    # exceptions
+    if data["decision"] < 1 or data["decision"] > 6:
+        return Response("Unsupported decision type", 400)
```
Comment (Astatium5) [inline src/api/events.py:511]: Should be `BadRequest` from `werkzeug.exceptions`

File: src/api/files.py [hunk 8]
```diff
@@ -188,12 +203,17 @@ def delete_file(file_id: str) -> Response:
     uid: str = decoded_token["uid"]
 
     # get file data from firestore
-    data = db.collection(u"Files").document(file_id)
+    data_ref = db.collection(u"Files").document(file_id)
+
+    if not data_ref.get().exists:
+        return Response("The file not found", 404)
```
Comment (Astatium5) [inline src/api/files.py:209]: Should be `NotFound` from `werkzeug.exceptions`

File: src/api/files.py [hunk 9]
```diff
@@ -208,16 +228,88 @@ def delete_file(file_id: str) -> Response:
...
+    author_uid: str = decoded_token["uid"]
+    data: dict = request.get_json()
+
+    # fetch the file data from firestore
+    file_ref = db.collection(u"Files").document(data["file_id"])
+    file = file_ref.get().to_dict()
+
+    # exceptions
+    if not file_ref.get().exists:
+        return Response(
```
Comment (Astatium5) [inline src/api/files.py:280]: Should be `NotFound` from `werkzeug.exceptions`

File: src/schemas.yml [hunk 10]
```diff
@@ -22,14 +45,30 @@ Event:
...
+    endtime:
+      type: timestamp
+      example: yy-MM-dd SSS HH:mm:ss
+    organizer:
+      type: string
+      example: The organier name
+    description:
+      type: string
+      example: Drink coffee with your teammates
+    participators:
```
Comment (Astatium5) [inline src/schemas.yml:69]: "participants"

File: src/schemas.yml [hunk 11]
```diff
@@ -1,18 +1,41 @@
...
+      type: timestamp
+      example: yy-MM-dd SSS HH:mm:ss
+    endtime:
+      type: timestamp
+      example: yy-MM-dd SSS HH:mm:ss
+    organizer:
       type: string
-      example: Your employer
-    eventDate:
+      example: The organier name
```
Comment (Astatium5) [inline src/schemas.yml:15]: "organizer"

File: src/api/files.py [hunk 12]
```diff
@@ -208,16 +228,88 @@ def delete_file(file_id: str) -> Response:
...
+    if "file" in data:
+        # save pdf to firestore storage
+        try:
+            bucket = storage.bucket()
+            blob = bucket.blob(data["file_id"])
+            blob.upload_from_string(
+                data["file"], content_type="application/pdf"
+            )
+        except:
+            raise BadRequest("cannot update to storage")
```
Comment (Astatium5) [inline src/api/files.py:307]: Looks like this is not a `BadRequest`, but maybe an `InternalServerError`. Not sure though, need to look at possible exceptions, which could be thrown by Storage

File: src/api/files.py [hunk 13]
```diff
@@ -188,12 +203,17 @@ def delete_file(file_id: str) -> Response:
...
+    if not data_ref.get().exists:
+        return Response("The file not found", 404)
+
+    data = data_ref.get().to_dict()
 
     # Only the author, reviewer, and admin have access to the data
     if (
-        uid != data.get("reviewer")
-        and uid != data.get("author")
+        uid != data["reviewer"]
```
Comment (Astatium5) [inline src/api/files.py:215]: Better stick with `.get()` for every retrieval from a dictionary operation. You can read more about it here (in the comparison section): *[link](https://www.programiz.com/python-programming/methods/dictionary/get)*

File: src/api/events.py [hunk 14]
```diff
@@ -221,16 +230,19 @@ def get_event() -> Response:
...
-            return jsonify(doc.to_dict()), 200
+    # fetch the event data from firestore
+    doc = db.collection("Scheduled-Events").document(event_id).get()
 
-        return Response(response="Event no longer exist", status=400)
-    except:
-        return Response(response="Failed to retrieve", status=400)
+    # event not found exception
+    if not doc.exists:
+        return Response(response="Event no longer exist", status=404)
```
Comment (Astatium5) [inline src/api/events.py:243]: Should be `NotFound` from `werkzeug.exceptions`

File: src/api/events.py [hunk 15]
```diff
@@ -143,55 +153,55 @@ def update_event() -> Response:
...
 
-        docs: list = (
-            db.collection("Scheduled-Events")
-            .where("id", "==", event_id)
-            .where("author", "==", uid)
-            .stream()
-        )
+    # if enevt does not exists
+    if not event_ref.get().exists:
+        return Response("The event not found", 404)
```
Comment (Astatium5) [inline src/api/events.py:181]: Should be `NotFound` from `werkzeug.exceptions`

File: src/api/events.py [hunk 16]
```diff
@@ -95,44 +103,46 @@ def delete_event(event_id: str) -> Response:
...
+    event = event_ref.get().to_dict()
 
-        # Delete document
-        for doc in docs:
-            # Store document id and use it to locate the specific document to delete
-            doc_id = doc.id
-            db.collection("Scheduled-Events").document(doc_id).delete()
+    # if enevt does not exists
+    if not event_ref.get().exists:
+        return Response("The event not found", 404)
```
Comment (Astatium5) [inline src/api/events.py:123]: Should be `NotFound` from `werkzeug.exceptions`

File: src/api/events.py [hunk 17]
```diff
@@ -112,15 +119,15 @@ def delete_event(event_id: str) -> Response:
...
 
     # fetch event data from firestore
     event_ref = db.collection("Scheduled-Events").document(event_id)
     event = event_ref.get().to_dict()
 
-    # if enevt does not exists
+    # if event does not exists
     if not event_ref.get().exists:
-        return Response("The event not found", 404)
+        return NotFound("The event not found", 404)
```
Comment (Astatium5) [inline src/api/events.py:130]: Not need to specify the response code, it is embedded in the exception itself. Same goes with every exception from `werkzeug.exceptions`

**Review by Astatium5 — CHANGES_REQUESTED** [review]

File: src/api/events.py (hunk 17 above)
Comment (TangyrenMidKing) [reply src/api/events.py:130]: I will change it asap, I need to go over the descriptions in the comments and change it.

**Review by Astatium5 — APPROVED** [review]

**Pull request merged**